databroker/muggler/data.py

```
@@ -119,19 +125,96 @@ class DataMuggler(object):
+            formatted like
+            [(<time>: {<data_key>: <value>, <data_key>: <value>, ...}), ...]
+        metatdata : dict
+            mapping data keys to source names
+
+            This information is used to look up resampling behavior.
+        """
+        raise NotImplementedError()
+        for event in event_tuples:
+            # TODO Make this look like an event object.
+            pass
 
-        self.max_frames = max_frames
-        self.recreate_columns(col_info)
+    @classmethod
+    def from_events(cls, events):
+        return cls(events)
+
+    def add_events(self, events):
+        self._stale = True
+        for event in events:
+            for name, description in event.descriptor.data_keys.items():
+
+                # If we have this source name, check for name collisions.
+                if name in self.sources:
+                    if self.sources[name] != description['source']:
+                        raise ValueError("In a previously loaded event, "
+                                         "'{0}' refers to {1} but in Event "
+                                         "{2} it refers to {3}.".format(
+                                             name, self.sources[name],
+                                             event.id,
+                                             description['source']))
+
+                # If it is a new name, determine a ColSpec.
+                else:
+                    self.sources[name] = description['source']
+                    if 'external' in event.descriptor.data_keys.keys():
+                        # TODO Figure out the specific dimension.
+                        pass
+                    else:
+                        dim = 0
+
+                    col_spec = ColSpec(name, dim, None, None)  # defaults
+                    # TODO Look up source-specific default in a config file
+                    # or some other source of reference data.
```

Allow the config file to override the default behavior. The hard-coded default must be the safest one: fail if someone tries to upsample data that has no referenced resampling behavior.